### ISSUE TITLE
Blazor Question: Order/Index of ChildComponents rendered via ChildContent/RenderFragment

### ISSUE THREAD

**Issue by DanielCordell** [dotnet/aspnetcore#54703]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Intentionally a duplicate of (closed) https://github.com/dotnet/aspnetcore/issues/20553 as it's basically the same use case, and I feel like that issue wasn't resolved. This isn't really a bug per-se, more of a question.

### Problem

There seems to be no "consistent" way of getting the render-order of items in a RenderFragment/ChildContent.

Lets say I have a component:
```html
<List>
  <ListItem Value="A"/>
  <ListItem Value="B"/>
  <ListItem Value="C"/>
</List> 
```

where it is crucial that I know the exact order of the `<ListItems>` as they appear in razor.

`<ListItem>` is a simple component like so:

```csharp
@{ Parent.TellParentMyOrder(this) }
@code
{
    [CascadingParameter]
    public List Parent { get; set; }
}
```

Now, the above component would work, as all three items are rendered at the same time. However, if I have a case like this:

```html
<List>
  <ListItem Value="A"/>
  @if(showItemTwo)
  { 
      <ListItem Value="B"/>
  }
  <ListItem Value="C"/>
</List>
```

In my initial render, `A` and `C` will render and tell the parent they exist. If I then set `showItemTwo` only `B` will render, so I then have the order `A`, `C`, `B`.

### Investigation into QuickGrid Columns

I thought about taking an approach similar to the way `QuickGrid` handles its columns, by "clearing" my list items each time `<List>` rerenders and then re-fetching them.
- You can see the exact behaviour I'm after on the [quickgrid docs](https://aspnet.github.io/quickgridsamples/columns) under `Dynamic Columns`. 
    - The "Name" column will always render to the left of the "Birth Date" column.
- However, as I mentioned before in my case, `A` and `C` won't rerender as their parameters haven't changed, leaving me with just a list of `B`.

I then thought I could call `StateHasChanged` from inside `<List>` on each saved item before I rerender, but I still end up with `A` and `C` rendering first before `B` (I presume as 'existing' components call BuildRenderTree before new ones, although I'm taking a guess there).

It seems like this only happens to work for `QuickGrid`'s dynamic column system because it just so happens that both provided column types require a reference type parameter that forces the component to rerender. It definitely looks like a side-effect, rather than an intentional design choice like I origianlly thought.

I created a new type of column for QuickGrid:

```CSharp
public class TestColumn<T> : ColumnBase<T>
{
    public override GridSort<T>? SortBy { get; set; }
    protected internal override void CellContent(RenderTreeBuilder builder, T item)
    {
        builder.AddContent(0, "HELLO");
    }
}
```

and this column when rendered conditionally will _always_ render at the end of the table, rather than visually where it should go.

### Expected Behavior

Ideally there could be some way of forcing the items to "render" in the order that they're written in (or just have this  be the default behaviour) but I accept this is a pretty niche use-case. I'm mostly just wondering if anyone has any ideas about something like this. 

I've gone through a few different files: https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/ChangeDetection.cs
https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs

But I can't really see any way I can force this sort of behaviour.

### Steps To Reproduce

The code in the issue I linked still applies https://github.com/thomasclaudiushuber/BlazorRenderFragmentPositionIssue/tree/master

Specifically the index page.
https://github.com/thomasclaudiushuber/BlazorRenderFragmentPositionIssue/blob/master/BlazorRenderFragmentPositionIssue/Pages/Index.razor

### .NET Version

8.0.102 but have tested on 6 as well.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. How this is done in QuickGrid (ensure you're passing an object-type parameter) is the only recommendation we would have. As for having a better / more first-class solution for this problem, we believe https://github.com/dotnet/aspnetcore/issues/17200 will be necessary, but that's currently in our Backlog.

**Comment by AscarGb** [comment]
> ### Is there an existing issue for this?
> * [x]  I have searched the existing issues
> 
> ### Describe the bug
> Intentionally a duplicate of (closed) #20553 as it's basically the same use case, and I feel like that issue wasn't resolved. This isn't really a bug per-se, more of a question.
> 
> ### Problem
> There seems to be no "consistent" way of getting the render-order of items in a RenderFragment/ChildContent.
> 
> Lets say I have a component:
> 
> ```
> <List>
>   <ListItem Value="A"/>
>   <ListItem Value="B"/>
>   <ListItem Value="C"/>
> </List> 
> ```
> 
> where it is crucial that I know the exact order of the `<ListItems>` as they appear in razor.
> 
> `<ListItem>` is a simple component like so:
> 
> ```cs
> @{ Parent.TellParentMyOrder(this) }
> @code
> {
>     [CascadingParameter]
>     public List Parent { get; set; }
> }
> ```
> 
> Now, the above component would work, as all three items are rendered at the same time. However, if I have a case like this:
> 
> ```
> <List>
>   <ListItem Value="A"/>
>   @if(showItemTwo)
>   { 
>       <ListItem Value="B"/>
>   }
>   <ListItem Value="C"/>
> </List>
> ```
> 
> In my initial render, `A` and `C` will render and tell the parent they exist. If I then set `showItemTwo` only `B` will render, so I then have the order `A`, `C`, `B`.

You can pass a delegate parameter, which will return the actual index of the element. And draw according to the order.

```
<List>
  <ListItem Value="A" GetIndex="() => 0"/>
  @if(showItemTwo)
  { 
      <ListItem Value="B" GetIndex="() => showItemTwo? 1 : -1"/>
  }
  <ListItem Value="C" GetIndex="() => showItemTwo? 2 : 1"/>
</List>
```

```
@code
{
    [CascadingParameter]
    public List Parent 
        {
            get => _parent;
            set
            {
                if (_parent!= value)
                {
                    _parent= value;
                    _parent.Add(this);
                }
            }
        }
}
```
